File: 028-字符串的排列/028-Permutation.py
# -*- coding:utf-8 -*-
class Solution:
    def __init__(self):
        self.ans=[]
    def Permutation(self, string):
        # write code here
        if not string:
            return []
        else:
            self.permute(string,0)
            return sorted(self.ans)
    def permute(self,string,begin):
        if begin==len(string)-1:
            self.ans.append(string)
        else:
            duplicate=set()
            for i in range(begin,len(string)):
                if string[i] not in duplicate:
                    duplicate.add(string[i])
                    newstr=string
                    if i!=begin:
                        newstr=string[:begin]+string[i]+string[begin+1:i]+string[begin]+string[i+1:]
                    self.permute(newstr,begin+1)
                    

File: 028-字符串的排列/028-Permutation.cpp
class Solution {
    vector<string> ans;
public:
    vector<string> Permutation(string str) {
        if(str.empty()){
            return ans;
        }
        permute(str,0);
        sort(ans.begin( ), ans.end( ));
        return ans;
    }
    void permute(string str,int begin){
        if(begin==str.size()){
            ans.push_back(str);
        }
        else{
            for(int i =begin;i<str.size();i++){
                if(!HasDuplicate(str,begin,i)){
                    swap(str[i],str[begin]);
                    permute(str,begin+1);
                    swap(str[i],str[begin]);
                }
            }
        }
    }
    bool HasDuplicate(string& str, int k, int i) const {
		for (int p = k; p < i; p++)
			if (str[p] == str[i]) return true;

		return false;
	}
};
